Source/Core/InputCommon/ControllerEmu/ControlGroup/IMUAccelerometer.cpp

```diff
@@ -0,0 +1,48 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "InputCommon/ControllerEmu/ControlGroup/IMUAccelerometer.h"
+
+#include "Common/Common.h"
+#include "Common/MathUtil.h"
+
+#include "Core/HW/WiimoteEmu/WiimoteEmu.h"
+
+#include "InputCommon/ControlReference/ControlReference.h"
+#include "InputCommon/ControllerEmu/Control/Control.h"
+#include "InputCommon/ControllerEmu/Control/Input.h"
+#include "InputCommon/ControllerEmu/ControllerEmu.h"
+#include "InputCommon/ControllerEmu/Setting/NumericSetting.h"
+
+namespace ControllerEmu
+{
+IMUAccelerometer::IMUAccelerometer(std::string name, std::string ui_name)
+    : ControlGroup(std::move(name), std::move(ui_name), GroupType::IMUAccelerometer),
+      m_has_data(false)
+{
+  controls.emplace_back(std::make_unique<Input>(Translate, _trans("Left")));
+  controls.emplace_back(std::make_unique<Input>(Translate, _trans("Right")));
+  controls.emplace_back(std::make_unique<Input>(Translate, _trans("Forward")));
+  controls.emplace_back(std::make_unique<Input>(Translate, _trans("Backward")));
+  controls.emplace_back(std::make_unique<Input>(Translate, _trans("Up")));
+  controls.emplace_back(std::make_unique<Input>(Translate, _trans("Down")));
+}
+
+std::optional<IMUAccelerometer::StateData> IMUAccelerometer::GetState()
+{
+  StateData state;
+  state.x = (controls[0]->control_ref->State() - controls[1]->control_ref->State());
+  state.y = (controls[3]->control_ref->State() - controls[2]->control_ref->State());
+  state.z = (controls[4]->control_ref->State() - controls[5]->control_ref->State());
+
+  if (state.x != 0 || state.y != 0 || state.z != 0)
```

Comment:
I don't like this check. Potentially a device in free-fall will return (0,0,0) and be interpreted as "no data".
Instead you can check if one of the `control_ref`s `BoundCount()` is non-zero.

That can also be done for the rotational velocity.

Reply:
> I don't like this check. Potentially a device in free-fall will return (0,0,0) and be interpreted as "no data".

That's why the check is done continously, and if anything but (0,0,0) is ever detected, the member variable m_has_data is set to `true` forever after.

Reply:
> Instead you can check if one of the control_refs BoundCount() is non-zero.

Does this return the number of inputs referenced in the control expression? If so, I can't see how that would help as the accelerometer and gyro mappings are, in the UDPServer case, intented to always contain their default values (as you see in the screenshot I recently uploaded).

Reply:
Oh. It's set to "true forever after". Gross. :P
I think it would be nicer to allow switching between "real" and "non-real" sources without hidden state in the background.
`BoundCount()` will give you the number of controls in the expression which are actually present. As long as your UDPServer doesn't add the device (and the inputs) to ControllerInterface until a DS4 is actually attached this should work.

Reply:
> BoundCount() will give you the number of controls in the expression which are actually present. As long as your UDPServer doesn't add the device (and the inputs) to ControllerInterface until a DS4 is actually attached this should work. The user would also have to choose the device in the drop-down for the "Yaw Left"-like expressions to be valid and "Bound".

That sounds perfect, I'll give it a try.